### PR TITLE
Bug 1874943 - Add a tab counter UI element

### PR DESCRIPTION
**Pull request by mavduevskiy**
I reused some of the classes already masterfully crafted by @MozillaNoah  for the nav bar as well :)

### Screenshots
 Old | New
 --- | --- 
 |  

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1874943

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (HarrisonOg) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:206]: why default to private?

Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:206]: Because it's a preview for the private theme, isn't it :)

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:101]: Why is this commented out and then in `NavigationBar` we use a `CompositionLocalProvider` for the same color?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: After @MatthewTighe's recent PR, you may be able to just directly observe the browsing mode from `AppState` (if that is ready to be consumed on this dev branch).

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:30]: nit:
```suggestion
 * @param browserStore [BrowserStore] used to observe the number of opened tabs.
```

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:64]: nit: there's a newer observe state function that returns a non-null value

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:55]: I might argue for this to be simplified:
```suggestion
    tabCount: Int,
```

As a leaf node in the Composable tree, `NavigationBar` doesn't need to have knowledge of the store here, and the mode is simply used as a logic control with the observation.

This will help simplify the previews, too.

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:106]:
```suggestion
                                    .size(48.dp),
```

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:119]: Is this preview compiling? This probably needs to get wrapped by `FirefoxTheme`

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:43]: are there any other changes to this file to be called-out from the original implementation?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: In fact, the `BrowsingModeManager` was removed completely so this change will be required once you rebase the dev branch onto main.

Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: Here is the patch for context https://github.com/mozilla-mobile/firefox-android/pull/4646

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (HarrisonOg) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:206]: d'oh! didn't see that

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: thanks! I think I will deal with it later on, I assume there will be more conflicts when I am about to merge this into main

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:64]: which one are you referring to?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:55]: I was thinking about the same thing, but I couldn't come up with a good way to organize the hierarchy. Where should we observe the store? Right in the `BottomToolbarContainerView`? I couldn't find any examples of observing things outside of some form of composable, that's hence I put it in the first composable we have.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:119]: Yes it does

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:43]: that's the comment I was waiting for. no, I kept the code exactly the same, except the location and naming!

**Comment by mavduevskiy** [comment]
@MozillaNoah @HarrisonOg 
thanks for the review, here goes another round :)

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:42]: Let's fix this warning, the ranges and numbers can be constants

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: Have yall been rebasing at all since making the dev branch?

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:109]: can you walk me through why this change was necessary? The removal of the color is going to break the coloring in the Tabs Tray

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:64]: https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/ext/ComposeExtensions.kt#62

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:55]: Yeah, in `BottomToolbarContainerView` makes sense to me. It would likely live there or further up, if that View were to ever be replaced with a Composable down the road.

There's a few examples of this with our ViewHolders:
https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt#45-61

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:92]: I'm not familiar with the tasks yall have already planned out for next steps. Will this click lambda eventually be used? Will the `IconButton` be used to trigger the showing of the Tabs Tray?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: I haven't yet

Comment (gabrielluong) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:42]: I think you might be doing disservice to yourself by not rebasing your dev branch with main and continuing forward with your current PR in its current state. Addressing merge conflicts before landing back to main does mean your original reviewed PRs will now contain changes that haven't been reviewed and will probably raise issues down the line compare to getting it right now. 

As an aside, I am wondering with this dev branch that we have agreed to use as rapid prototyping to reach some findings on how we should proceed with implementing this project - has there been any discovery or decision reached on how we should proceed? Eg, was this gonna to be done in XML or a mixed of XML/Compose where possible? 

If y'all believe there is a good path forward and there are any share outs that you can provide the team to help validate that this is a good way to move forward. I would recommend we move towards landing your changes back onto main if the investigative work has been productive and has support from the team.

If we are still prototyping, then feel free to continue with the development branch.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:109]: I am not changing the colour, it's exactly the way it is in the original [implementation](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayBanner.kt#367) :)

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:92]: I am not sure yet what's the best approach to implement clicks. Do you have any suggestions?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:45]: @gabrielluong 
That's a very good point, [here](https://docs.google.com/document/d/1Lo76-tYhkgB-8LZh5nefaOxouIPFTo_x2x8w2_VPbhU/edit) it is. I was actually going to ask you to take a look. I don't intend to wait for the whole team approval, but everybody is welcomed to make suggestions/raise concerns. 

I was actually approaching the dev branch as something that eventually will become a patch to merge into main.

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt]: added for consistency

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:110]: My bad; I totally missed the content color assignment in `TabsTrayBanner`. We're good here 👍

File: fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt:41]: nit: Since this will be in its own file (and to avoid any coloring confusion in the future), could you just add a small note to the kdoc that the icon and text are colored via `LocalContentColor` and `CompositionLocalProvider`?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:75]: An `IconButton` should be fine, I was just asking to clarify if the click was actually going to be consumed, and, if so, what the follow-up was going to be.

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:45]: Before I do my final pass, what is the verdict here? When will the dev branch be going away?

I don't have anything more to add about the perils of not rebasing a dev branch with main periodically; GL's comment summarized my thoughts/concerns.

Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:45]: I get the concerns. My intention was to open up a PR for a review, to merge the dev branch into the main. It's not the most optimal way to do it, but it would have a benefit of commits being previously reviewed so it won't have surprises during the final review. I was going to rebase before this final feature version.

I did rebase it now to speed up the process, but ironically Matt's change has been reverted.

Comment (gabrielluong) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:45]: I think it's worthwhile to highlight some risks with how we approach with the project and areas where we want to make sure we are using the best judgement:

- Given this would've been a multi-month development branch, it would be highly encouraged that we continue to rebase the dev branch to avoid bitrot and leaving rebase and merge conflicts at the end. This was probably the most notable red flag here. This isn't a point against using the development branch, but more about how we are  mitigating risk in our day-to-day development.
- This particular patch has some upstream implications because it is essentially refactoring the tab counter UI that is in-use today. I think it would be better to consider landing the part of the refactoring on main and then rebasing the change back into the development branch. This will require some judgement on your part in your development cycle because ideally we keep the isolated feature inside of the development branch. Think of the development branch as where you can prototype quickly, keep a feature in complete secret until it's ready to merge and ship or if you're working on a redesign/change of an existing feature that cannot be feature flag easily because it builds on top of something existing. The point is the development branch is to keep all the feature development separate and isolated from main. Given how the current PR can be beneficial upstream, I would highly encourage landing this upstream to make it easier for everyone and rebasing the development branch with your changes.
- When we initially discussed the development branch, we were still trying to prototype quickly between using Compose and XML or both, and ensure we only landed what was needed in main at the end. How ya'll been reviewing and landing incremental changes into the development branch to build the new feature has been what we wanted to see. Looking at your doc, I think the next steps would be to communicate the implementation details and get buy-in from a domain expert on the browser toolbar in terms of how we are implementing this. I don't really know if I am the domain expert here, but I also recognize that such a person might not exist on the team anymore. The likely candidate I can think of is Jon or Roger.
- There is a couple of things that have changed with this project since we started and I think it's time to start thinking about how we can mitigate some of these risks in the coming month. One reasoning for this is that your primary reviewer on this project will be going on leave soon. Any new reviewer will not have the context of the previous patches that will be landing back into the main, which will cause its own set of issues. Another concern is that the monorepo migration will likely happen soon. Timing wise I think Noah provides a good call out for re-considering when we should merge back into main. It already looks like the development branch has reached a big milestone if it is clear to y'all how the implementation needs to be approached and you can get buy-in from someone who is familiar with the existing browser toolbar.

My proposed next steps:

1. Let's land the tab counter UI refactoring into main and then we can have a different PR that uses this new element in the dev branch.
2. Find a domain expert (this could be 1-3 people, not everyone needs to be an expert, but ideally one person is familiar with the toolbar) around the browser toolbar and start communicating your implementation plan. Aim to find a reviewer who can take over next month in this process.
3. Communicate a plan on how you want to proceed with the dev branch and the pathway back to main. We can absolutely continue with the dev branch in the short term and nothing needs to change. By Feb, we will need to have an idea of how we will mitigate the concerns that I have listed above. My recommendation is to aim to reland this back in main earlier once we have established that the implementation plan is sound. An alternative idea would be to plan for an appropriate milestone checkpoint (maybe when the feature no longer needs the incomplete feature flag) as when you decide to reland back to main. You will need to work out the timing of this because this involves people's schedule for leave, how long you think this need to be in development for, and monorepo migration timeline.

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:75]: just to close the loop, synced with @MozillaNoah about it; the tap will be triggering the same behaviour we currently have in the toolbar when the tab button is clicked.

**Comment by mavduevskiy** [comment]
converting this to a draft until a separate patch with compose [refactoring](https://github.com/mozilla-mobile/firefox-android/pull/5365) lands

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:45]: We had a follow up with @gabrielluong, takeaways are
1. [landed](https://github.com/mozilla-mobile/firefox-android/pull/5365) refactoring separately 
2. gl will be taking a closer look into the project, effectively taking a domain expert role. We lost Jonathan, who had a lot of context about the toolbar, to another team, the navigation bar change doesn't affect the toolbar itself a lot, but rather is adding new compose elements while reusing the existing code,- given all of that I am very happy to see him helping out with it.
3. I am planning to open a PR with the most feature next week, which should give us enough time before Harrison going on parental leave and monorepo migration. In other words, I expect to merge a few smaller patches for dev branch, and then the feature is ready for main.

**Comment by mavduevskiy** [comment]
@HarrisonOg I think we have green light now! could you give it a final look?
pinging all the participants for visibility

**Review by HarrisonOg — APPROVED** [review]
@mavduevskiy I think there's something weird with the commits in this PR but other than that, lgtm!

**Review by MozillaNoah — APPROVED** [review]
LGTM for this iteration! Thanks for bearing with the review churn here

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:47]: We want to put the required parameters before the optional parameters. For the most part, you will see we follow https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-component-api-guidelines.md#parameters-order. The except we implement is that lambdas can appear below optional parameters.

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:74]: I am wondering why we need this?

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:52]: You can also get the browserStore via `components.core.store` since `components` is exposed in @composable.

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:74]: We need to use `LocalContentColor` because of its original usage in `TabsTray`. The tab page buttons are colored via this property, which, under the hood, uses a combination of a color and an alpha to color the selected/unselected states.

Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:74]: https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayBanner.kt#211-261

It's a nuance of `TabRow` and `Tab` (the UI elements; not in a browser context).

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt]: thanks, did that

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/BottomToolbarContainerView.kt:52]: changed to `observeAsState(initialValue = 0) `, the parameter name is slightly different

File: fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/NavigationBar.kt:74]: some explanation exists in the class [itself](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/compose/TabCounter.kt#43)

**Pull request merged**
